internal/cephfs/nodeserver.go

```diff
@@ -127,10 +133,55 @@ func maybeUnlockFileEncryption(
 	stagingTargetPath string,
 	volID fsutil.VolumeID,
 ) error {
-	if volOptions.IsEncrypted() {
-		log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
 
-		return fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
+	if !volOptions.IsEncrypted() {
+		return nil
+	}
+
+	// Define Mutex Lock variables
+	lockName := string(volID) + "-mutexLock"
+	lockCookie := lockName + "-coockie"
+	lockDesc := "Lock for " + string(volID)
+	lockDuration := 150 * time.Second
+	var flags byte = 0
+
+	log.DebugLog(ctx, "Creating lock for the following volume ID %s", volID)
+
+	ioctx, err := volOptions.GetConnection().GetIoctx(volOptions.MetadataPool)
+	if err != nil {
+		log.ErrorLog(ctx, "Failed to create ioctx: %s", err)
+	}
+	defer ioctx.Destroy()
+
+	res, err := ioctx.LockExclusive(volOptions.VolID, lockName, lockCookie, lockDesc, lockDuration, &flags)
+	switch res {
+	case 0:
+		log.DebugLog(ctx, "Lock successfully created for volume ID %s", volID)
+	case -C.EBUSY:
+		log.DebugLog(ctx, "Lock is already held by another client and cookie pair")
+	case -C.EEXIST:
+		log.DebugLog(ctx, "Lock is already held by the same client and cookie pair")
+	default:
+		return fmt.Errorf("Failed to lock volume ID %v: %w", volID, err)
+	}
+
+	if res == 0 {
+		log.DebugLog(ctx, "cephfs: unlocking fscrypt on volume %q path %s", volID, stagingTargetPath)
+		err = fscrypt.Unlock(ctx, volOptions.Encryption, stagingTargetPath, string(volID))
+		if err != nil {
+			return err
+		} else {
```

Comment:
no need for the `else {`, just continue with the function and loose one indention level (makes it easier to read and understand the code).